Split chat lines into pools and allow the player to filter them

dragunoff opened the pull request:
There's a lot going on here so I've tried to split it across commits in a logical way:
- Split chat lines into pools (e.g. System, Mission, Chat, Notifications, UI Feedback)
- Allow the player to toggle the display of certain chat pools (e.g. hide UI Feedback text)
- Add transcription support to speech notifications (e.g. display a chat line when EVA says "Unit lost")
- Add transcriptions to speech notifications in TD (add transcriptions to relevant EVA quotes)

Here are some screenshots to explain the idea:
![ingame-chat](https://user-images.githubusercontent.com/1355810/75615819-9d6f0a00-5b51-11ea-9ce0-9a141cee0869.png)
*In-game chat with filter options and transcriptions*

![display-settings](https://user-images.githubusercontent.com/1355810/75615839-c5f70400-5b51-11ea-9eca-ade734c482d8.png)
*Display settings*

![ingame-chat-common](https://user-images.githubusercontent.com/1355810/75615844-d4ddb680-5b51-11ea-92f8-885d73c407c0.png)
*Filter options for mods without transcriptions*

Depends on #17663 (for the custom dropdown arrow image). The in-game chat filter panel is not necessary but I think it's a nice addition - bringing the filters in context and allowing the player to quickly change the settings. I had to add the cog glyph to D2k and TS - and ArtSrc PR will follow if this gets approved.

Closes #16148
Closes #6400
Closes #6354
Perhaps #3278 as well?
Provides the basis for solving #5775 and #12421

dragunoff committed:
Split chat lines into pools

- Add a common class for passing around chat lines
- Add wrapper methods for adding chat lines
- Combine repeated chat lines in the display widget

dragunoff committed:
Allow the player to toggle the display of certain chat pools

dragunoff committed:
Add transcription support to speech notifications

dragunoff committed:
Add transcriptions to speech notifications in TD

dragunoff commented:
Rebased.

netnazgul commented:
I'd suggest adding a short hint description of each option as it might not be entirely clear what is covered by which checkbox.

abcdefg30 commented:
Needs a rebase again. :/

dragunoff commented:
Rebased. Confilct were only in the chrome image files (because of merging #17806).

dragunoff commented:
Rebased after the merge of #17663

dragunoff commented:
Update: rebased after the merge of #18366

pchote commented:
What else is needed for this to fully solve #5775 and #12421?

I haven't had a chance to test this yet, but the images above suggest that spectators can already turn chat off?

dragunoff commented:
@pchote The UI in the screenshots is a bit misleading perhaps. I included "Chat", "Mission" and "System" as disabled checkboxes so that it makes sense that the user is seeing stuff in the chat even when all available boxes are off. 

Filtering out spectator chat is made possible by this PR but there are some caveats with the UI. We need to disable the spec chat for spectators that have filtered it out - i.e. they should only have access to the "All" chat. This won't be a problem in replays.

If we want to allow filtering out the main chat as well, then we need to ensure that the user still has access to the chat window - for typing commands.

And there is an important distinction to be made between filtering chat messages and hiding the chat overlay.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs:
```diff
@@ -227,44 +234,64 @@ public void OpenChat()
+					chatLine.Pool);
+			}
+			else
+			{
```
Style nit: Omit braces.

dragunoff replied:
FIxed.

abcdefg30 commented on mods/cnc/audio/notifications.yaml:
```diff
@@ -4,49 +4,76 @@ Speech:
 		ConstructionComplete: constru1
+			Transcription: Construction complete.
 		EnemyUnitsApproaching: enmyunit
+			Transcription: Enemy approaching.
```
`enmyunit` actually says "Enemy unit destroyed." Seems like you transcribed `enmyapp1`?

pchote replied:
The announcement name is `EnemyUnitsApproaching` so this is just highlighting an earlier mistake?

dragunoff replied:
`EnemyUnitsApproaching` is not used anywhere and probably that's why it was not spotted until now. Should I take any action here?

abcdefg30 reviewed (dismissed):
The concept lgtm ingame. One thing I wondered about were transcriptions and (more importantly) UI feedback selectors for spectators. I'm not sure if exposing that makes sense for spectators, but on the other hand I'm not sure if it is worth special casing something.

dragunoff commented:
> 
> 
> The concept lgtm ingame. One thing I wondered about were transcriptions and (more importantly) UI feedback selectors for spectators. I'm not sure if exposing that makes sense for spectators, but on the other hand I'm not sure if it is worth special casing something.

Currently UI feedback includes only the messages coming from selection hotkeys ("Selected across screen / map") and these are relevant for spectators. They don't hear the announcer but I'd agree it's not worth special casing this. And fundamentally these are user settings that are also found in the settings panel and live in `settings.yaml` so it makes sense that they be the same regardless of player status.

abcdefg30 commented:
Ok, I'm fine with that, so just "fixup requested" for the older remarks.

dragunoff commented on OpenRA.Game/Game.cs:
```diff
@@ -863,14 +863,36 @@ public static void AddSystemLine(string text)
-		public static void AddSystemLine(string name, string text)
+		public static void AddSystemLine(string prefix, string text)
 		{
-			OrderManager.AddChatLine(name, systemMessageColor, text, systemMessageColor);
+			AddChatLine(prefix, systemMessageColor, text, systemMessageColor, ChatPool.System);
 		}
 
 		public static void AddChatLine(string name, Color nameColor, string text)
 		{
-			OrderManager.AddChatLine(name, nameColor, text, chatMessageColor);
+			AddChatLine(name, nameColor, text, chatMessageColor, ChatPool.Chat);
+		}
+
+		public static void AddChatLine(string name, Color nameColor, string text, Color textColor,  ChatPool pool)
+		{
+			OrderManager.AddChatLine(new ChatLine(name, nameColor, text, textColor, pool));
+		}
+
+		public static void AddTranscribedChatLine(string text)
+		{
+			if (Settings.Game.ChatPoolFilters.HasFlag(ChatPoolFilters.Transcriptions))
+				AddChatLine("Battlefield Control", systemMessageColor, text, systemMessageColor, ChatPool.Transcriptions);
+		}
+
+		public static void AddFeedbackChatLine(string text)
+		{
+			if (Settings.Game.ChatPoolFilters.HasFlag(ChatPoolFilters.Feedback))
+				AddChatLine("Battlefield Control", systemMessageColor, text, systemMessageColor, ChatPool.Feedback);
+		}
+
+		public static void AddMissionChatLine(string name, Color nameColor, string text)
+		{
+			AddChatLine(name, nameColor, text, chatMessageColor, ChatPool.Mission);
```
One thing that bothers me with this implementation is that now there are a bunch of add-chat-line methods cluttering `Game.cs`. I thought about moving them to their own file but I couldn't figure out how to do it. IIRC it had something to do with passing the `OrderManager` instance. I didn't dig too deep into that though. And of course this can be done in a follow-up.

dragunoff commented:
Update: fixed style nit.

dragunoff commented:
Update: Rebased and resolved conflicts.

pchote commented:
I like the idea of the audio transcripts, but I'm sorry to say that I still find the UI presentation confusing and a bit of a deal breaker.

I understand the motivation for the ingame filter list and think this will be good UI once we actually have the ability to filter standard and team/spectator chat. But IMO right now it feels premature, and I expect the disabled options to cause a lot of confusion. I'm also not sure we should allow players to disable mission/system messages at all (and what is the difference between "System" and "UI feedback"?).

Likewise for the settings menu, where my confusion was amplified further by testing this PR in RA which doesn't have a "Transcriptions" option at all! Having to click a dropdown to access only one usable checkbox feels quite wrong.

Does anybody have a list of references for how other games handle this?

pchote commented:
IMO it would be clearer if we could have a single "Show audio notifications in chat" checkbox in the audio settings, and extract the "Selected across screen" etc messages into notifications.yaml so they fall under the same setting. The ingame filtering could then be limited to all/team chat?

dragunoff commented:
> 
> 
> I like the idea of the audio transcripts, but I'm sorry to say that I still find the UI presentation confusing and a bit of a deal breaker.
> 
> I understand the motivation for the ingame filter list and think this will be good UI once we actually have the ability to filter standard and team/spectator chat. But IMO right now it feels premature, and I expect the disabled options to cause a lot of confusion. I'm also not sure we should allow players to disable mission/system messages at all (and what is the difference between "System" and "UI feedback"?).
> 
> Likewise for the settings menu, where my confusion was amplified further by testing this PR in RA which doesn't have a "Transcriptions" option at all! Having to click a dropdown to access only one usable checkbox feels quite wrong.
> 

Mission and system messages can't be disabled. 

"UI Feedback" currently is used only for _"Selected across map/screen"_ messages - i.e. UI events that don't give any other visual or audible feedback. The "System" pool is for events such as game start/end, player connects/disconnets, etc. and for debug.

I added the disabled checkboxes for completeness (each checkbox corresponds to a chat pool) and to make it clear to the player why they are seeing stuff in chat even when they've unchecked all boxes. 

Initially I added a "Spectators" pools for toggling spectator chat but decided to leave it off for now as its a bit trickier with the UI for that feature.

> 
> 
> IMO it would be clearer if we could have a single "Show audio notifications in chat" checkbox in the audio settings, and extract the "Selected across screen" etc messages into notifications.yaml so they fall under the same setting. The ingame filtering could then be limited to all/team chat?

_"Selected across..."_ messages are not audible so they are not the same thing (well, there is a beep sound but no voice line). Lumping them together with the narrator's lines would make it impossible to see these notifications without seeing also all of the narrator's lines. Which may be OK... Perhaps other should chime in here and we should have a look at how other games do it.

pchote commented:
In any case this isn't critical for the next release, so dropping from the next release as part of a wider cleanup to try and unstall progress there.

Mailaender reviewed (approved):
Works as promised and code looks good.

dragunoff commented:
I still very much have interest in finishing this. I'd be fine with removing the controversial UI and leaving that for a follow-up when we can also implement a spectator chat filter.

dragunoff commented:
@pchote How about something like this for the UI:
![image](https://user-images.githubusercontent.com/1355810/99888501-081f2300-2c56-11eb-9cf2-a88cbc36f5f0.png)

pchote commented:
How about putting them all on the same line?

<img width="626" alt="Screenshot 2020-11-21 at 22 50 01" src="https://user-images.githubusercontent.com/167819/99889337-fe90bd80-2c4b-11eb-9ed0-791d7b93ef50.png">

This would avoid breaking the y spacing between each row.

FWIW I still think it would make sense for this setting to live on the Audio tab instead.

dragunoff commented:
@pchote I don't mind putting them on line line, will do. To me it makes more sense to put it in the "Display" tab because it is really about things displaying on the screen. Or perhaps it's time to seriously consider overhauling the settings panel to support arbitrary amount of settings and tabs (not in this PR of course). I'd be interested to work on that for next+1.

dragunoff commented:
Closing in favor of #18905
